src/main/java/ch/njol/skript/util/ClassInfoReference.java

```diff
@@ -0,0 +1,158 @@
+		return Kleenean.get(isPlural);
+	}
+
+	/**
+	 * Wraps a ClassInfo expression as a ClassInfoReference expression which will return ClassInfoReferences
+	 * with as much derivable context as possible
+	 * @param classInfoExpression the ClassInfo expression to wrap
+	 * @return a wrapper ClassInfoReference expression
+	 */
+	@NonNull
+	public static Expression<ClassInfoReference> wrap(@NonNull Expression<ClassInfo<?>> classInfoExpression) {
+		if (classInfoExpression instanceof ExpressionList) {
+			ExpressionList<?> classInfoExpressionList = (ExpressionList<?>) classInfoExpression;
+			Expression<ClassInfoReference>[] wrappedExpressions = Arrays.stream(classInfoExpressionList.getExpressions())
+				.map(expression -> wrap((Expression<ClassInfo<?>>) expression))
+				.toArray(Expression[]::new);
+			return new ExpressionList<>(wrappedExpressions, ClassInfoReference.class, classInfoExpression.getAnd());
+		}
+		Kleenean isPlural = determineIfPlural(classInfoExpression);
+		if (classInfoExpression instanceof Literal) {
+			Literal<ClassInfo<?>> classInfoLiteral = (Literal<ClassInfo<?>>) classInfoExpression;
+			ClassInfo<?> classInfo = classInfoLiteral.getSingle();
+			return new SimpleLiteral<>(new ClassInfoReference(classInfo, isPlural), classInfoLiteral.isDefault());
+		}
+		return new SimpleExpression<ClassInfoReference>() {
+
+			@Override
+			protected @Nullable ClassInfoReference[] get(Event event) {
+				if (classInfoExpression.isSingle()) {
+					ClassInfo<?> classInfo = classInfoExpression.getSingle(event);
+					if (classInfo == null) {
+						return new ClassInfoReference[0];
+					}
+					return new ClassInfoReference[] { new ClassInfoReference(classInfo, isPlural) };
+				}
+				return classInfoExpression.stream(event)
+					.map(ClassInfoReference::new)
+					.toArray(ClassInfoReference[]::new);
+			}
+
+			@Override
+			public boolean isSingle() {
+				return classInfoExpression.isSingle();
+			}
+
+			@Override
+			public Class<? extends ClassInfoReference> getReturnType() {
+				return ClassInfoReference.class;
+			}
+
+			@Override
+			public String toString(@Nullable Event event, boolean debug) {
+				if (debug) {
+					return classInfoExpression.toString(event, true) + "(wrapped by " + getClass().getSimpleName() + ")";
+				}
+				return classInfoExpression.toString(event, false);
+			}
+
+			@Override
+			public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult) {
+				return classInfoExpression.init(expressions, matchedPattern, isDelayed, parseResult);
+			}
+		};
+	}
+
+	private Kleenean plural;
+	private ClassInfo<?> classInfo;
+
+	public ClassInfoReference(ClassInfo<?> classInfo) {
+		this.plural = Kleenean.UNKNOWN;
+		this.classInfo = classInfo;
+	}
+
+    public ClassInfoReference(ClassInfo<?> classInfo, Kleenean plural) {
+        this.plural = plural;
+        this.classInfo = classInfo;
+    }
```

Comment:
```suggestion
	private ClassInfoReference(ClassInfo<?> classInfo) {
		this(classInfo, Kleenean.UNKNOWN);
	}

    private ClassInfoReference(ClassInfo<?> classInfo, Kleenean plural) {
        this.plural = plural;
        this.classInfo = classInfo;
    }
```
Would we want to enforce usage of the `wrap` method?

Reply:
I'd like to keep it public in case you need an instance without an expression